pkg/agent/nplagent/rules/rules.go

```diff
@@ -0,0 +1,57 @@
+// +build !windows
+
+// Copyright 2020 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rules
+
+import (
+	"github.com/coreos/go-iptables/iptables"
+	nplutils "github.com/vmware-tanzu/antrea/pkg/agent/nplagent/lib"
+)
+
+type PodPortRules interface {
+	Init() (bool, error)
+	AddRule(port int, podip string) (bool, error)
+	DeleteRule(port int, podip string) (bool, error)
+	SyncState(podPort map[int]string) bool
+	GetAllRules(podPort map[int]string) (bool, error)
+	DeleteAllRules() (bool, error)
+}
+
+func Initrules(args ...string) PodPortRules {
+	var ruletype string
+	if len(args) == 0 {
+		// By default use iptable
+		ruletype = nplutils.IPTableRule
+	} else {
+		ruletype = args[0]
+	}
+	switch ruletype {
+	case "IPTABLE":
+		ipt, err := iptables.New()
+		if err != nil {
+			return nil
+		}
+		iptRule := IPTableRule{
+			name:  "NPL",
+			table: ipt,
+		}
+
+		iptRule.DeleteAllRules()
+		iptRule.CreateChains()
+		return &iptRule
+	}
+	return nil
+}
```

Comment:
maybe I'm missing something, but this seems to really defeat the point of having an interface.

there should be a `NewIPTableRule` constructor and the client should be responsible for calling `Init` on the returned object, which would in turn take care of creating new chains, etc if necessary.

Reply:
Ideally we should be able to figure out the type of implementation for PodPortRules from the underneath system's capability or configuration specified. In this case, as we are only supporting one kind of implementation we don't need to do that.
Though, I get your point about adding an abstraction. I'll update it.